Update CSRF/antiforgery docs for deferred validation behavior (#37262)
dotnet/AspNetCore.Docs: aspnetcore/security/csrf-protection.md

monikerRange: '>= aspnetcore-11.0'
ms.author: tdykstra
ms.custom: mvc
ms.date: 06/17/2026
uid: security/csrf-protection
---
# Automatic CSRF protection in ASP.NET Core

> Starting in .NET 11, ASP.NET Core ships an automatic Cross-Site Request Forgery (CSRF) protection middleware that's enabled by default in apps built with `WebApplication.CreateBuilder`. Unlike the [token-based antiforgery system](xref:security/anti-request-forgery), this middleware doesn't issue or validate tokens. Instead, it inspects the [Fetch Metadata](https://developer.mozilla.org/docs/Web/HTTP/Headers/Sec-Fetch-Site) headers that modern browsers attach to every request and the [`Origin`](https://developer.mozilla.org/docs/Web/HTTP/Headers/Origin) header as a fallback, then records a validation verdict on the request. Components that process submitted form data enforce that verdict, rejecting cross-origin write requests that aren't explicitly trusted.

Reviewer: ```suggestion
Starting in .NET 11, ASP.NET Core ships an automatic Cross-Site Request Forgery (CSRF) protection middleware that's enabled by default in apps built with `WebApplication.CreateBuilder`. Unlike the [token-based antiforgery system](xref:security/anti-request-forgery), this middleware doesn't issue or validate tokens. Instead, it inspects the [Fetch Metadata](https://developer.mozilla.org/docs/Web/HTTP/Headers/Sec-Fetch-Site) headers that modern browsers attach to every request, with the [`Origin`](https://developer.mozilla.org/docs/Web/HTTP/Headers/Origin) header as a fallback, then records a validation verdict on the request. Components that process submitted form data enforce that verdict, rejecting cross-origin form posts that aren't explicitly trusted.
```
Very minor, you can take or leave this.  Updating the intro to be more consistant with the rest of the article's details and the way it refers to form posts.

1. **Any other `Sec-Fetch-Site` value is denied.** When `Sec-Fetch-Site` is `cross-site` or `same-site` and the origin isn't trusted via CORS, the request is denied.
1. **No `Sec-Fetch-Site`, but `Origin` is present:** the middleware compares the `Origin` to `scheme://host[:port]` built from the request. If they match, the request is allowed; otherwise it's denied. This is the fallback path for browsers older than the Fetch Metadata spec (released ~2020).
1. **No `Sec-Fetch-Site` and no `Origin`: the request is allowed.** Browsers always send at least one of these on a write request, so a request missing both is almost certainly a non-browser client such as `curl`, Postman, a mobile app, or a server-to-server caller. CSRF is a browser-only attack vector, so these requests pass through.

> The middleware records this verdict on the request rather than ending the request itself. How and when a denied verdict turns into an HTTP `400 Bad Request` response is covered in [Deferred validation](#deferred-validation).

Reviewer: ```suggestion
The middleware records this verdict on the request rather than ending the request itself. For how and when a denied verdict turns into an HTTP `400 Bad Request` response, see [Deferred validation](#deferred-validation).
```
Very minor: "is covered in" was a bit awkward.

When this key is set to `true`, `WebApplication` skips registering the middleware in the pipeline. The `ICsrfProtection` service remains registered, so anything that resolves it directly continues to work.

> > [!NOTE]

Reviewer: ```suggestion
> [!WARNING]
```
Consider treating this as a warning rather than a note, as it is presented elsewhere here.

This is a behavior change for Blazor SSR apps that previously removed `app.UseAntiforgery()`: they're now protected by the CSRF middleware rather than left unprotected, and they stop emitting antiforgery tokens. For migration guidance, see <xref:migration/antiforgery-to-csrf>. For the formal breaking-change notice, see [Blazor server-side rendering defers antiforgery validation to middleware](/aspnet/core/breaking-changes/11/blazor-server-side-rendering-deferred-cross-site-request-forgery-protection).

## Adopting CSRF-only protection in existing apps

For apps that are upgrading from .NET 10 and already use the token-based system, the automatic CSRF middleware can replace it entirely in many scenarios. For guidance on when to drop the token-based system and how to do so, see <xref:migration/antiforgery-to-csrf>.

## Troubleshooting

**Symptom:** Same-origin requests from a browser succeed, but cross-origin form posts return `400` with no body.

> **Cause:** The CSRF middleware recorded an invalid verdict for the cross-origin request, and a form-processing component — an MVC action, a minimal API form binding, or a Blazor SSR form post — enforced that verdict with a `400`. This is the expected default behavior for endpoints that process forms.

Reviewer: ```suggestion
**Cause:** The CSRF middleware recorded an invalid verdict for the cross-origin request, and a form-processing component — such as an MVC action, a Minimal API form binding, or a Blazor SSR form post — enforced that verdict with a `400`. This is the expected default behavior for endpoints that process forms.
```
Very minor item:  See if this wording is a little better.  Also Minimal API is treated as a proper noun in this article elsewhere as it should, matching it here.
